src/WebJobs.Extensions.DocumentDB/Bindings/DocumentDBAttributeBindingProvider.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host;
+using Microsoft.Azure.WebJobs.Host.Bindings;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DocumentDB
+{
+    internal class DocumentDBAttributeBindingProvider : IBindingProvider
+    {
+        private JobHostConfiguration _jobHostConfig;
+        private DocumentDBConfiguration _docDBConfig;
+
+        public DocumentDBAttributeBindingProvider(JobHostConfiguration config, DocumentDBConfiguration documentDBConfig)
+        {
+            _jobHostConfig = config;
+            _docDBConfig = documentDBConfig;
+        }
+
+        public async Task<IBinding> TryCreateAsync(BindingProviderContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            ParameterInfo parameter = context.Parameter;
+            DocumentDBAttribute attribute = parameter.GetCustomAttribute<DocumentDBAttribute>(inherit: false);
+            if (attribute == null)
+            {
+                return null;
+            }
+
+            if (_docDBConfig.EndpointUri == null)
+            {
+                throw new InvalidOperationException(
+                    string.Format(CultureInfo.CurrentCulture,
+                    "The DocumentDB Endpoint Uri must be set either via a '{0}' app setting or directly in code via DocumentDBConfiguration.EndpointUri.",
+                    DocumentDBConfiguration.AzureWebJobsDocumentDBUriName));
+            }
+
+            if (string.IsNullOrEmpty(_docDBConfig.AuthKey))
+            {
+                throw new InvalidOperationException(
+                    string.Format(CultureInfo.CurrentCulture,
+                    "The DocumentDB Authentication Key must be set either via a '{0}' app setting or directly in code via DocumentDBConfiguration.AuthKey.",
+                    DocumentDBConfiguration.AzureWebJobsDocumentDBKeyName));
+            }
+
+            DocumentDBContext documentDBContext = CreateContext(_docDBConfig, attribute, _jobHostConfig.NameResolver);
+
+            await CreateIfNotExists(attribute, documentDBContext);
+
+            Type coreType = DocumentDBUtility.GetCoreType(parameter.ParameterType);
+
+            IBinding binding = null;
+            try
+            {
+                binding = BindingFactory.BindGenericCollector(parameter, typeof(DocumentDBAsyncCollector<>), coreType,
+                _jobHostConfig.GetService<IConverterManager>(), (s) => documentDBContext);
+            }
+            catch (Exception ex)
+            {
+                // BindingFactory uses a couple levels of reflection so we get TargetInvocationExceptions here.
```

Comment:
Exceptions from BindingFactory should be easier to handle. Filed issue in sdk repo to track: https://github.com/Azure/azure-webjobs-sdk/issues/652